### PR TITLE
Move the New UI / front-end into a stand-alone top-level directory

### PR DESCRIPTION
**Pull request by edalex-ian**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This moves the New UI / front-end react/typescript project into it's own top level directory in the project. It used to be deeply nested down in the SBT/Java structure @ `Source/Plugins/Core/com.equella.core/js` whereas now it resides in `react-front-end`.

This was done to

1. Make it more readily accessible - both for day to day development, and for potential community contribution; and
2. As a step towards building this in a true stand-alone module (similar to the REST Module) - albeit this is one step of many required.

Main areas of work:

1. Splitting out the Jest test for the version checking - now in its own NPM project at `Source/Plugins/Core/com.equella.core/test/javascript`;
2. Move the code to `react-front-end` and tweak various NPM related scripts;
3. Tweak the SBT setup to then pull back in what it needs - changes to `./build.sbt` and `Source/Plugins/Core/com.equella.core/build.sbt` as well as `project/CommonSettings.scala`; and
4. Updating CI scripts to ensure all is still built and tested.

Note, some changes to `Source/Plugins/Core/com.equella.core/build.sbt` were just to address SBT deprecation warnings while there.

I'm initially doing this as a draft PR so that I can check the CI side of things; what that all works then it looks like I'll need to merge `develop` in due to various `package.json` updates with the old file structure - that'll be fun.

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by edalex-ian** [comment]
Well that's good, the CodeBuild worked - and that's running against New UI. Now to try and merge in all the renovate updates.

**Comment by edalex-ian** [comment]
Wow, git pretty well automatically merged that in. Let's see how it goes though.

**Comment by edalex-ian** [comment]
This is ready for review now @PenghaiZhang :wink:

File: Source/Plugins/Core/com.equella.core/test/javascript/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+module.exports = {
+  preset: "ts-jest",
+  testMatch: ["**/?(*.)+(spec|test).[jt]s?(x)"],
+};
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/test/javascript/jest.config.js:21]: this is added to support the version check test?

**Review by ChristianMurphy — APPROVED** [review]
Exciting to see this come together.
I'm not as familiar with the SBT part, but the JS/TS part looks good.
Thanks @edalex-ian!

File: Source/Plugins/Core/com.equella.core/test/javascript/jest.config.js (hunk 1 above)
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/test/javascript/jest.config.js:21]: Yup, that's correct. There's a now a clean split between the testing of the old stuff and new stuff.

**Comment by edalex-ian** [comment]
Gah, looks like more renovate goodies. I'll do another merge from `develop` but then once @PenghaiZhang is happy I'll just merge to `develop` - as we've now had several green builds and renovate would probably cause another merge conflict before I get there.

File: build.sbt [hunk 2]
```diff
@@ -228,6 +228,24 @@ writeScriptingJavadoc := {
...
+ThisBuild / reactFrontEndOutputDir := reactFrontEndDir.value / "target/resources"
+ThisBuild / buildReactFrontEnd := {
+  val dir = reactFrontEndDir.value
+  Common.nodeInstall(dir)
+  Common.nodeScript("build", dir)
+
+  // return the location of the resulting artefacts
+  reactFrontEndOutputDir.value
+}
+ThisBuild / reactFrontEndLanguageBundle := reactFrontEndOutputDir.value / "lang/jsbundle.json"
```
Comment (PenghaiZhang) [inline build.sbt:241]: These tasks and settings are mostly used in the core sbt but defined here. I guess the reason is react front-end is now on the root level ?

**Review by PenghaiZhang — APPROVED** [review]
Thanks Ian. LGTM!

**Comment by edalex-ian** [comment]
Excellent, thank you both. Merging now before renovate adds more.

**Pull request merged**